{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"This notebook is designed to retrieve 16s data from One Codex and perform differential abundance analysis between cohorts of mice treated\n",
    "with various drug products over a timecourse.  \"\"\"\n",
    "\"\"\"Importing modules\"\"\"\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import argparse\n",
    "from onecodex import Api\n",
    "import functools\n",
    "import seaborn as sns\n",
    "import json\n",
    "# import ecopy as ep\n",
    "from skbio.stats.composition import ancom\n",
    "from skbio.stats.ordination import pcoa\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA as sklearnPCA\n",
    "from matplotlib import pyplot as plt\n",
    "from sklearn.metrics import pairwise_distances\n",
    "import matplotlib\n",
    "from matplotlib.colors import ListedColormap\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.collections import PatchCollection\n",
    "%matplotlib inline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Setting Box directory\"\"\"\n",
    "data_dir = '/Users/home/data/'\n",
    "results_dir = '/Users/home/results/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"\"\"Finding samples\"\"\"\n",
    "ocx = Api('API-KEY')\n",
    "all_samples = ocx.Samples.all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Accessing samples by project\"\"\"\n",
    "project = ocx.Projects.get('PROJECT_ID')\n",
    "samples = ocx.Samples.where(project=project.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"checking that correct samples have been pulled and metadata is complete\"\"\"\n",
    "print(len(samples))\n",
    "one_sample = ocx.Samples.get('ONE_SAMPLE_ID')\n",
    "print(one_sample.metadata.custom)\n",
    "print(one_sample.metadata.name)\n",
    "\"\"\"Looking at metadata columns present in One Codex Data\"\"\"\n",
    "samples.metadata[['external_sample_id', 'description', 'drug']]\n",
    "\"\"\"\n",
    "external_sample_id = subject id\n",
    "description = timepoint\n",
    "drug = treatment\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Getting Shannon Diversity and generating boxplots based on timepoint and treatment\"\"\"\n",
    "shannon_diversity = samples.alpha_diversity('shannon')\n",
    "shannon_diversity = shannon_diversity.merge(samples.metadata[['external_sample_id', 'description', 'drug']], left_index=True, \n",
    "                                           right_index=True)\n",
    "sns.set(context='talk', style='darkgrid')\n",
    "def shannon_plots(shannon_frame):\n",
    "    \n",
    "    plot_df = shannon_frame\n",
    "    plot_df.sort_values(by='description', inplace=True, ascending=False)\n",
    "    \"\"\"Setting order of timepoints\"\"\"\n",
    "    category_order = ['time_x', 'time_y', 'time_z']\n",
    "    print(plot_df.head())\n",
    "    g = sns.catplot(y='shannon', x='description', data=plot_df[['shannon', 'description', 'drug']], hue='drug', \n",
    "               kind='box', height=6, aspect=6/4, hue_order=['trt_1', 'trt_2', 'trt_3', 'trt_4', 'trt_5'],\n",
    "                   order=category_order)\n",
    "    g.axes[0,0].set(ylim=(0,6))\n",
    "    g.set_xticklabels(['Day x', 'Day y', 'Day z'])\n",
    "    g.fig.suptitle(str('Shannon Diversity'))\n",
    "    g.axes[0,0].set_xlabel('Time Point')\n",
    "    g.axes[0,0].set_ylabel('Shannon Diversity')\n",
    "    g.fig.subplots_adjust(top=.9)\n",
    "    g._legend.set_title('Treatment')\n",
    "    plt.savefig(results_dir + 'plots/Shannon_Diversity_box.png')\n",
    "\n",
    "\n",
    "shannon_plots(shannon_diversity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples.plot_mds(method='pcoa', metric='braycurtis', color='drug')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Running 2 way anova with post hoc tukey test\"\"\"\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "model = ols('shannon ~ C(description) + C(drug) + C(description):C(drug)', data=shannon_diversity).fit()\n",
    "sm.stats.anova_lm(model, typ=2)\n",
    "shannon_diversity['combination'] = shannon_diversity.description + \" / \" + shannon_diversity.drug\n",
    "\n",
    "m_comp = pairwise_tukeyhsd(endog=shannon_diversity['shannon'], groups=shannon_diversity['combination'], alpha=0.05)\n",
    "\n",
    "tukey_data = pd.DataFrame(data=m_comp._results_table.data[1:], columns = m_comp._results_table.data[0])\n",
    "\n",
    "sig_results = tukey_data[tukey_data['reject'] == True]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Defining functions to create the tables that can be used in ANCOM and as a summary output\"\"\"\n",
    "\n",
    "\"\"\"Putting the readcount tables into a dictionary and filtering. Filters on minimum readcount and taxonomic rank.  \n",
    "Genus or Family is appropriate for this 16S method.\"\"\"\n",
    "def get_sample_dfs(sample_list, min_readcount, rank):\n",
    "    sample_dict = {}\n",
    "    for i in range(len(sample_list)):\n",
    "        \"\"\"Need to define the name for this sample\"\"\"\n",
    "        sample_id = sample_list[i].metadata.name.replace('.fastq.gz', '')\n",
    "        \n",
    "        readcount_table = sample_list[i].primary_classification.table()[['name', 'rank', 'readcount_w_children']]\n",
    "        readcount_table = readcount_table[readcount_table['rank'] == rank]\n",
    "        readcount_table = readcount_table[readcount_table['readcount_w_children'] >= min_readcount]\n",
    "        readcount_table = readcount_table[['name', 'readcount_w_children']]\n",
    "        \"\"\"ANCOM needs the readcount column to be renamed to the sample it belongs to\"\"\"\n",
    "        readcount_table.columns = ['classification', sample_id]\n",
    "        readcount_table.set_index('classification', inplace=True)\n",
    "        sample_dict[sample_id] = readcount_table\n",
    "    return sample_dict\n",
    "\n",
    "\"\"\"Merging readcount dfs from the dictionary.  Used to create df for ANCOM (need to remove missing numbers before running ANCOM)\"\"\"\n",
    "def dataframe_dict_merge(dict_dataframes):\n",
    "    key_list = list(dict_dataframes.keys())    \n",
    "    readcount_dfs = [dict_dataframes[x] for x in key_list]\n",
    "    merged_function_dfs = functools.reduce(lambda left,right: pd.merge(left,right, right_index=True, left_index=True, \n",
    "                                                                      how='outer'), readcount_dfs)\n",
    "    merged_function_dfs.columns = key_list\n",
    "    merged_function_dfs.fillna(1, inplace=True)\n",
    "    ancom_df = merged_function_dfs.copy()\n",
    "    \n",
    "    \"\"\"Creates a multi index by splitting the name strings.\"\"\"\n",
    "    idx = merged_function_dfs.columns.str.rsplit('_', expand=True)\n",
    "    merged_function_dfs.columns = idx\n",
    "    merged_function_dfs.sort_index(level=0, axis=1, inplace=True)\n",
    "    return merged_function_dfs, ancom_df\n",
    "\n",
    "\"\"\"Collapse readcounts on sample and get percent composition for samples\"\"\"\n",
    "def count_by_sample(merged_frame):\n",
    "    \"\"\"getting sample names which should be level 0\"\"\"\n",
    "    sample_names = list(merged_frame.columns.levels[0])\n",
    "    \"\"\"allocating df\"\"\"\n",
    "    all_counts = pd.DataFrame()\n",
    "    for i in sample_names:\n",
    "        sample_count = merged_frame.groupby(level=0, axis=1).sum()[i].sort_values(ascending=False).reset_index()\n",
    "        sample_count['percentage'] = (sample_count[i] / sample_count[i].sum()) * 100\n",
    "        sample_count.columns = [str(i + '_Classification'), str(i + '_ReadCount'), str(i + '_PercentComposition')]\n",
    "        sample_count = sample_count.head(20)\n",
    "        all_counts = pd.concat([sample_count, all_counts], axis=1)\n",
    "    idx = all_counts.columns.str.split('_', expand=True)\n",
    "    all_counts.columns=idx\n",
    "    all_counts.sort_index(level=0, axis=1, inplace=True)\n",
    "\n",
    "    return all_counts\n",
    "\n",
    "\n",
    "\"\"\"Present percent composition by timepoints for all samples\"\"\"\n",
    "def count_by_timepoint(merged_frame):\n",
    "    \"\"\"get sample name and timepoints\"\"\"\n",
    "    sample_names = merged_frame.columns.levels[0]\n",
    "    time_points = merged_frame.columns.levels[1]\n",
    "    #print(time_points)\n",
    "    \"\"\"allocating df\"\"\"\n",
    "    all_counts = pd.DataFrame()\n",
    "    for i in sample_names:\n",
    "        for k in time_points:\n",
    "            sample_count = merged_frame[i][k].sort_values(ascending=False).reset_index()\n",
    "            sample_count['percentage'] = (sample_count[k] / sample_count[k].sum()) * 100\n",
    "            sample_count.columns = [str(i + '_' + k + '_Classification'), str(i + '_' + k + '_ReadCount'), \n",
    "                                    str(i + '_' + k + '_PercentComposition')]\n",
    "            sample_count = sample_count.head(20)\n",
    "            #print(x)\n",
    "            all_counts = pd.concat([sample_count, all_counts], axis=1)\n",
    "\n",
    "    idx = all_counts.columns.str.split('_', expand=True)\n",
    "    all_counts.columns=idx\n",
    "    all_counts.sort_index(level=0, axis=1, inplace=True)\n",
    "    return all_counts\n",
    "\n",
    "            \n",
    "merged_frame, ancom_df = dataframe_dict_merge(get_sample_dfs(samples, 10, 'genus'))\n",
    "time_point_count_frame = count_by_timepoint(merged_frame)\n",
    "sample_count_frame = count_by_sample(merged_frame)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"ANCOM Calculations, creating multiple groupings for later in the analysis.  Need a df specifically for ANCOM\n",
    "ANCOM does not like 1 counts, specifiy a minimum of 10 in the ancom generating function\"\"\"\n",
    "time_grouping = pd.Series(list(x.metadata.description for x in samples), \n",
    "                     index=list(x.metadata.name.replace('.fastq.gz', '') for x in samples))\n",
    "sample_grouping = pd.Series(list(x.metadata.external_sample_id for x in samples), \n",
    "                     index=list(x.metadata.name.replace('.fastq.gz', '') for x in samples))\n",
    "drug_grouping = pd.Series(list(x.metadata.custom['drug'] for x in samples),\n",
    "                         index=list(x.metadata.name.replace('.fastq.gz', '') for x in samples))\n",
    "\n",
    "\n",
    "results = ancom(ancom_df.transpose(), drug_grouping)\n",
    "\n",
    "\n",
    "results[results['W'] > 0]\n",
    "\"\"\"Alternate ways of selecting for differentially abundant genera\"\"\"\n",
    "bacterial_list = results[results['reject'] == True].index.tolist()\n",
    "bacterial_list = results[results['W'] > 5].index.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Setting up ancom metadata frame for downstream pivot table and possible multiple comparisons\"\"\"\n",
    "time_grouping_df = pd.DataFrame(time_grouping)\n",
    "sample_df = pd.DataFrame(sample_grouping)\n",
    "drug_df = pd.DataFrame(drug_grouping)\n",
    "ancom_metadata = pd.merge(time_grouping_df, sample_df, left_index=True, right_index=True)\n",
    "ancom_metadata = pd.merge(ancom_metadata, drug_df, left_index=True, right_index=True)\n",
    "ancom_metadata.columns = ['description', 'external_sample_id', 'drug']\n",
    "ancom_metadata['category'] = 'bacteria'\n",
    "ancom_metadata.loc[ancom_metadata['drug'] == 'vehicle', ['category']] = 'vehicle'\n",
    "ancom_metadata.loc[ancom_metadata['drug'] == 'naïve', ['category']] = 'naïve'\n",
    "# ancom_metadata.head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ancom(ancom_df, metadata, comparison):\n",
    "    \"\"\"comparison should be either drug or category comparison\"\"\"    \n",
    "    \"\"\"comparision dict for between comparators at the three timepoints\"\"\"\n",
    "    comparator_dict = {}\n",
    "    \"\"\"comparison dict for within comparator timepoints\"\"\"\n",
    "    timepoint_dict = {}\n",
    "#     ancom_results_dict = {}\n",
    "    comparison_list = metadata[comparison].unique().tolist()\n",
    "    \n",
    "    \"\"\"Getting between comparator ancom for the all timepoints\"\"\"\n",
    "    for i in ['time_x', 'time_y', 'time_z']:\n",
    "        print(i)\n",
    "        subset_meta = metadata[metadata['description'] == i].copy()\n",
    "        print(len(subset_meta))\n",
    "        grouping = pd.Series(subset_meta[comparison])\n",
    "        print(len(grouping))\n",
    "        column_list = grouping.index.tolist()\n",
    "        subset_ancom = ancom_df[column_list]\n",
    "        print(len(subset_ancom.T))\n",
    "        name_string = str(comparison + '_' + str(i))\n",
    "        print(name_string)\n",
    "        comparator_dict[name_string] = ancom(subset_ancom.transpose(), grouping)\n",
    "    \n",
    "    for i in comparison_list:\n",
    "        \"\"\"Within category comaprison for the three timepoints\"\"\"\n",
    "        \"\"\"Getting the metadata columns for comparison, and then using these to pull out\n",
    "        the proper columns from the larger ancom df to run ancom on again\"\"\"\n",
    "        subset_meta = metadata[metadata[comparison] == i]\n",
    "        grouping = pd.Series(subset_meta['description'])\n",
    "        column_list = grouping.index.tolist()\n",
    "        subset_ancom = ancom_df[column_list]\n",
    "        name_string = str('within_' + str(i))\n",
    "        print(name_string)\n",
    "        \"\"\"Saving Ancom results to dictionary\"\"\"\n",
    "        timepoint_dict[name_string] = ancom(subset_ancom.transpose(), grouping)\n",
    "                \n",
    "        \n",
    "\n",
    "    return comparator_dict, timepoint_dict\n",
    "\n",
    "\n",
    "comparator_dict, timepoint_dict = get_ancom(ancom_df, ancom_metadata, 'drug')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\"\"\"Using ANCOM results generate barplots of significantly differentiated genera for both within timepoint comparisons of treatment, and \n",
    "withing sample comparisons accross timepoint\"\"\"\n",
    "\n",
    "def get_ancom_summary(comparator_dict, timepoint_dict, ancom_df, metadata, comparison):\n",
    "    \n",
    "    \"\"\"Dicts of results where ancom is true and the sig genera summarized by grouping\"\"\"\n",
    "    comparator_summary = {}\n",
    "    timepoint_summary = {}\n",
    "    time_point_list = metadata['description'].unique().tolist()\n",
    "    comparison_list = metadata[comparison].unique().tolist()\n",
    "    \"\"\"Pivot ancom df into right format\"\"\"\n",
    "    merged_ancom = ancom_df.T.merge(ancom_metadata[['description', 'external_sample_id', 'drug', \n",
    "                                                    'category']], left_index=True, right_index=True, how='left')\n",
    "    pivoted = merged_ancom.pivot_table(index=['drug', 'description']).T\n",
    "    pivoted = (pivoted / pivoted.sum(axis=0)) * 100\n",
    "    \n",
    "    \"\"\"Getting within timepoint comparisons for differentiated genera between treatments\"\"\"\n",
    "    for i in time_point_list: \n",
    "        day = ''\n",
    "        results_frame = comparator_dict[str(comparison + '_' + i)]\n",
    "        sig_genera = results_frame[results_frame['reject'] == True].index.tolist()\n",
    "        print(i, len(sig_genera))\n",
    "        sig_frame = pivoted.loc[sig_genera]\n",
    "        comparator_summary[(str(comparison + '_' + i))] = sig_frame.loc[:, sig_frame.columns.get_level_values(1).isin([i])]\n",
    "    \n",
    "    \"\"\"Getting within sample diff across timepoints\"\"\"\n",
    "    for i in comparison_list:\n",
    "        day = ''\n",
    "        results_frame = timepoint_dict[str('within_' + i)]\n",
    "        sig_genera = results_frame[results_frame['reject'] == True].index.tolist()\n",
    "        print(i, len(sig_genera))\n",
    "        sig_frame = pivoted.loc[sig_genera]\n",
    "        timepoint_summary[(str('within_' + i))] = sig_frame.loc[:, sig_frame.columns.get_level_values(0).isin([i])]\n",
    "#         print(sig_genera)\n",
    "#         print(timepoint_summary[(str('within_' + i))])\n",
    "        ax = timepoint_summary[(str('within_' + i))].T.plot(kind='bar', stacked=True, figsize=(12,10), \n",
    "                                                                    edgecolor='none', cmap='tab20b')\n",
    "\n",
    "        ax.set_xticklabels(['Day X',  \n",
    "                            'Day Y',  \n",
    "                            'Day Z', ], rotation = 45, ha='right')\n",
    "        ax.legend(bbox_to_anchor=(1,0.9), title='Genus')\n",
    "\n",
    "        ax.set_ylabel('Percent Composition')\n",
    "        ax.set_xlabel('Date')\n",
    "        ax.set_title('Significantly Differentiated Genera within ' + i + ' Timepoints')\n",
    "        ax.set_ylim(0,5)    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(results_dir + 'plots/sig_diff_genera_' + i + '_timepoints.png')\n",
    "    return comparator_summary, timepoint_summary\n",
    "comparator_results, timepoint_results = get_ancom_summary(comparator_dict, timepoint_dict, ancom_df, ancom_metadata, 'drug')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Getting top genera by cohort, dose, timepoint using pivot table\"\"\"\n",
    "\n",
    "ancom_meta_merged = ancom_df.T.merge(ancom_metadata[['description', 'external_sample_id', 'drug', 'category']], left_index=True, right_index=True, how='left')\n",
    "pivoted = ancom_meta_merged.pivot_table(index=['drug', 'description']).T\n",
    "pivoted = (pivoted / pivoted.sum(axis=0)) * 100\n",
    "comp_frame = pd.DataFrame()\n",
    "for col in pivoted.columns.sort_values(ascending=False):\n",
    "    temp_frame = pd.DataFrame(pivoted[col].nlargest(20).reset_index())\n",
    "    comp_frame = comp_frame.merge(temp_frame, left_index=True, right_index=True, how='right')\n",
    "\n",
    "    \n",
    "comp_frame.to_excel(results_dir + 'tables/all_samples_sumary_top20.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Plotting top ten most abundant genera by drug and timepoing\"\"\"\n",
    "colorblind_hex = ['#4477AA', '#332288', '#6699CC', '#88CCEE', '#44AA99', '#117733',\n",
    "            '#999933', '#DDCC77', '#661100', '#CC6677', '#AA4466', '#882255',\n",
    "            '#AA4499']\n",
    "cblind_cmap = matplotlib.colors.ListedColormap(colorblind_hex, name='my_colormap_name')\n",
    "\n",
    "\n",
    "\"\"\"Identifying top 10 genera by all samples\"\"\"\n",
    "cohort_pivot = pivoted.copy()\n",
    "cohort_pivot['sum'] = cohort_pivot.sum(axis=1)\n",
    "cohort_pivot.sort_values(by='sum', inplace=True, ascending=False)\n",
    "top_ten = cohort_pivot.head(10).copy()\n",
    "top_ten\n",
    "# top_ten['description']\n",
    "top_ten.drop(['sum'], axis=1, inplace=True)\n",
    "\n",
    "col_names = [' '.join(col).strip() for col in top_ten.columns.values]\n",
    "top_ten.columns = top_ten.columns.droplevel(0)\n",
    "top_ten.columns = col_names\n",
    "index_list = top_ten.index.to_list()\n",
    "\n",
    "\"\"\"Getting amount for remaining genera\"\"\"\n",
    "other_values = {}\n",
    "for i in col_names:\n",
    "    other_values[i] = 100 - sum(top_ten[i])\n",
    "top_ten = top_ten.append(other_values, ignore_index=True)\n",
    "index_list.append('Other Genera')\n",
    "top_ten.index = index_list\n",
    "\n",
    "\"\"\"Stacked bar plot of top ten genera composition\"\"\"\n",
    "ax = top_ten.T.plot(kind='bar', stacked=True, figsize=(12,10), edgecolor='none', cmap='tab20b')\n",
    "\n",
    "ax.set_xticklabels(['trt_1 Day_x','trt_1 Day_y', 'trt_1 Day_z',\n",
    "                    'trt_2 Day_x', 'trt_2 Day_y', 'trt_2 Day_z',\n",
    "                    'trt_3 Day_x', 'trt_3 Day_y', 'trt_3 Day_z', \n",
    "                    'trt_4 Day_x', 'trt_4 Day_y', 'trt_4 Day_z', \n",
    "                    'trt_5 Day_x', 'trt_5 Day_y', 'trt_5 Day_z'], rotation = 45, ha='right')\n",
    "ax.legend(bbox_to_anchor=(1,0.9), title='Genus')\n",
    "\n",
    "ax.set_ylabel('Percent Composition')\n",
    "ax.set_xlabel('Date and Treatment')\n",
    "ax.set_title('Top Ten Genera Percent Composition By 16S Sequencing')\n",
    "ax.set_ylim(0,110)\n",
    "plt.tight_layout()\n",
    "plt.savefig(results_dir + 'plots/genera_comp_all_conditions_stacked_barplot.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Getting a distance matrix for all of the samples from the OCX function and calculating the PCOA using skbio\n",
    "These functions are separate from the above functions, and so rank is specified here again\"\"\"\n",
    "def run_pcoa(ocx_sample_list):\n",
    "\n",
    "    \"\"\"getting filenames to rename columns based on the OCX id\"\"\"\n",
    "    id_to_filename_dict = {}\n",
    "    distance_matrix = ocx_sample_list._compute_distance(rank='genus', metric='braycurtis').to_data_frame()\n",
    "    \n",
    "    # Perform PCOA\n",
    "    ord_result = pcoa(distance_matrix.round(6))\n",
    "    plot_data = ord_result.samples.iloc[:, [0, 1]]\n",
    "    plot_data = plot_data.div(plot_data.abs().max(axis=0), axis=1)\n",
    "    plot_data.index = distance_matrix.index\n",
    "    plot_data = plot_data.rename(index=lambda x: ocx.Classifications.get(x).sample.metadata.name)\n",
    "    plot_data = plot_data.rename(index=lambda x: x.replace('_fastq.gz', ''))\n",
    "    plot_data = plot_data.rename(index=lambda x: x.replace('.fastq.gz', ''))\n",
    "    return plot_data\n",
    "\n",
    "pco_data = run_pcoa(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Getting plotting data and generating plot for pco by treatment and timepoint\"\"\"\n",
    "pco_plotting = pco_data.merge(ancom_metadata, left_index=True, right_index=True, how='left')\n",
    "\n",
    "pco_plotting.rename(columns={'description':'TimePoint','drug':'Treatment'}, inplace=True)\n",
    "cohort_df.sort_values(by=['external_sample_id', 'Date'], inplace=True)\n",
    "fig, ax = plt.subplots(figsize=(25,17))\n",
    "sns.scatterplot(cohort_df['PC1'], cohort_df['PC2'], hue=cohort_df['Treatment'], style=cohort_df['TimePoint'], s=300)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=1, title='Treatment and TimePoint', prop={'size':20})\n",
    "plt.title('16S Microbiome Profiling PCoA by Treatment and Timepoint')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(results_dir + 'plots/all_pcoa.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Creating radial bar plots by individual subject for each timepoint with top 25 genera\"\"\"\n",
    "def get_radial_plotting_df(ancom_df):    \n",
    "    \"\"\"get and sort by totals\"\"\"\n",
    "    ancom_df['totals'] = ancom_df.sum(axis=1)\n",
    "    ancom_df.index.name = 'genus'\n",
    "    ancom_df.sort_values(ascending=False, by='totals', inplace=True)\n",
    "    \"\"\"Convert to percentages\"\"\"\n",
    "    cols = ancom_df.columns.tolist()\n",
    "    percent_frame = ancom_df[cols].div(ancom_df[cols].sum(axis=0), axis=1).multiply(100).copy()\n",
    "    percent_frame.reset_index(inplace=True)\n",
    "    top_25_df = percent_frame.head(25)\n",
    "    \"\"\"calculate theta values\"\"\"\n",
    "    top_25_df['theta'] = (2 * np.pi / len(top_25_df['genus'])) * top_25_df.index\n",
    "    return top_25_df\n",
    "\n",
    "radial_plotting_df = get_radial_plotting_df(ancom_df)\n",
    "subject_list = samples.metadata[['external_sample_id']].unique().astype(str).tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_radials(subject_list, dataframe, start, color_map):\n",
    "    \"\"\"setting global parameters\"\"\"\n",
    "    sns.set(style='white', context='talk')\n",
    "    width = .3\n",
    "    bottom = 20\n",
    "    \"\"\"by adjusting the first item in the linespace call, we can change where on the colormap things start. However things must \n",
    "    stay between 0 and 1, else everything outside that range will be the same color\"\"\"\n",
    "    color_idx = np.linspace(start, 1, 25)\n",
    "    \"\"\"Different colormap options\"\"\"\n",
    "#     my_cmap = plt.cm.get_cmap('nipy_spectral')\n",
    "#     my_cmap = plt.cm.get_cmap('tab20c')\n",
    "#     my_cmap = plt.cm.get_cmap('gist_ncar')\n",
    "#     my_cmap = plt.cm.get_cmap('hsv')\n",
    "#     my_cmap = plt.cm.get_cmap('gist_rainbow')\n",
    "    \"\"\"Seaborn colormaps\"\"\"\n",
    "#     my_cmap = ListedColormap(sns.color_palette('Spectral_r', n_colors=25, desat=.8))\n",
    "    my_cmap = plt.cm.get_cmap(color_map)\n",
    "    colors = my_cmap(color_idx)\n",
    "    timepoints = ['x', 'y', 'z']\n",
    "\n",
    "    \"\"\"looping through subjects and saving individual figures to one final pdf\"\"\"\n",
    "    pdf = matplotlib.backends.backend_pdf.PdfPages(results_dir + '16s_combined_radials.pdf')\n",
    "    for subject_id in subject_list:\n",
    "        legend_list = dataframe['genus']\n",
    "        fig = plt.figure(figsize=(15,10)) # Notice the equal aspect ratio\n",
    "        ax1, ax2, ax3 = [fig.add_subplot(3,3,i+1, polar=True) for i in range(3)]\n",
    "        #plt.figure(figsize=(20,10))\n",
    "        #ax1 = plt.subplot(331, polar=True)\n",
    "        #ax2 = plt.subplot(332, polar=True)\n",
    "        #ax3 = plt.subplot(333, polar=True)\n",
    "        axes = [ax1, ax2, ax3]\n",
    "        gs1 = gridspec.GridSpec(1, 3)\n",
    "        gs1.update(wspace=0.0, hspace=0.0)\n",
    "        \"\"\"batching axes setup\"\"\"\n",
    "        for ax in axes:\n",
    "            ax.set_theta_direction(-1)\n",
    "            ax.set_theta_zero_location('NE')\n",
    "            ax.set_rlabel_position(340)\n",
    "            #ax.set_rgrids(labels=['', '10', '20', '30'], rotation='vertical')\n",
    "            #ax.set_rgrids([10, 30, 40, 50], labels=['', '10', '20', '30'], rotation='vertical')\n",
    "            \n",
    "\n",
    "            ax.tick_params(labelsize=10, rotation='auto')\n",
    "            ax.set_thetagrids([], labels=[])\n",
    "            ax.spines['polar'].set_visible(False)\n",
    "        for i in range(len(timepoints)):       \n",
    "            for x,y,c,lb in zip(dataframe['theta'], dataframe[str(subject_id + timepoints[i])], colors, legend_list):\n",
    "                axes[i].bar(x, y, color=c,label=lb, width=.25, bottom=50, linewidth=0)\n",
    "            axes[i].set_rgrids([60, 70, 80, 90], labels=['10', '20', '30', '40'], rotation='vertical')\n",
    "            axes[i].set_aspect('equal')\n",
    "        \n",
    "        \"\"\"Creating Global Legend\"\"\"\n",
    "        ax1.legend(loc=(1.43, 1.1), ncol=5, prop={'size': 8})\n",
    "        ax1.set_title(str('Subject ' + subject_id + '\\n 16S Microbiome Profile'), y=1.1)\n",
    "        ax1.set_xlabel('Day X', y=-.2)\n",
    "        ax2.set_xlabel('Day Y', y=-.2)\n",
    "        ax3.set_xlabel('Day Z', y=-.2)\n",
    "        #ax3.set_xlabel('Washout')\n",
    "        plt.tight_layout()\n",
    "        plt.subplots_adjust(wspace=.0, hspace=0)\n",
    "        plt.savefig(results_dir + 'Subject_' + subject_id + '_radial_bar.png', dpi=300, bbox_inches='tight', pad_inches=.5)\n",
    "        pdf.savefig()\n",
    "    pdf.close()\n",
    "    \n",
    "plot_radials(subject_list, radial_plotting_df, 0.00, 'nipy_spectral')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
